# Player Trueskill

### Helpful Resources
[Trueskill Documentation](https://trueskill.org/#version-0-4-4)  
[Trueskill for DOTA2 Players - Kaggle](https://www.kaggle.com/devinanzelmo/dota-2-skill-rating-with-trueskill)  
[Collection of Trueskill Examples](https://python.hotexamples.com/examples/trueskill/TrueSkill/-/python-trueskill-class-examples.html)  

In [1]:
# Housekeeping
import requests
import datetime
import pandas as pd
import numpy as np
import io
import os
import trueskill
import itertools
import math

import oeutils

pd.options.display.max_rows = 100
pd.options.display.max_columns = 500

workingdir = '../data/'
years = ['2018', '2019', '2020', '2021']

#oeutils.download(workingdir, years)
df = oeutils.read(workingdir, years)
df = oeutils.clean(df, split_on='player', keep_identities=False, keep_leagues=False, keep_columns=False)

/home/paulperrone/source/lol-analysis/notebooks/oeutils.py:160: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  oe_data['gameid'] = oe_data['gameid'].str.strip()


In [2]:
df.shape

(188865, 117)

In [3]:
df['league'].unique()

array(['LPL', 'LCK', 'CK', 'VCS', 'LCS.A', 'EU LCS', 'LLN', 'OPL', 'LMS',
       'TCL', 'CBLOL', 'CLS', 'NA LCS', 'TCS', 'LJL', 'LCL', 'CIS CL',
       'BRCC', 'GPL', 'EM', 'MSI', 'OCS', 'WCS', 'UL', 'NASG', 'DC',
       'KeSPA', 'LEC', 'LLA', 'TRA', 'BIG', 'LFL', 'LCS', 'UKLC', 'LDL',
       'UPL', 'PCS', 'Riot', 'MSC', 'RCL', 'CU', 'NLC', 'NEST', 'BL',
       'PGN', 'LCK CL', 'GLL', 'SL', 'PRM', 'CBLOL.A', 'EBL', 'BM'],
      dtype=object)

In [4]:
df.head()

gameid datacompleteness                                          url  \
348  2899-3157         complete  https://lpl.qq.com/es/stats.shtml?bmid=2899   
349  2899-3157         complete  https://lpl.qq.com/es/stats.shtml?bmid=2899   
350  2899-3157         complete  https://lpl.qq.com/es/stats.shtml?bmid=2899   
351  2899-3157         complete  https://lpl.qq.com/es/stats.shtml?bmid=2899   
352  2899-3157         complete  https://lpl.qq.com/es/stats.shtml?bmid=2899   

    league  year   split  playoffs                date  game  patch  playerid  \
348    LPL  2018  Spring         0 2018-01-15 09:40:00     1   8.01         1   
349    LPL  2018  Spring         0 2018-01-15 09:40:00     1   8.01         2   
350    LPL  2018  Spring         0 2018-01-15 09:40:00     1   8.01         3   
351    LPL  2018  Spring         0 2018-01-15 09:40:00     1   8.01         4   
352    LPL  2018  Spring         0 2018-01-15 09:40:00     1   8.01         5   

     side position      player             team    champion  ban1      ban2  \
348  Blue      top        Duke  Invictus Gaming        Ornn  Azir  Malzahar   
349  Blue      jng        Ning  Invictus Gaming     Kha'Zix  Azir  Malzahar   
350  Blue      mid      Rookie  Invictus Gaming     Orianna  Azir  Malzahar   
351  Blue      bot  JackeyLove  Invictus Gaming      Ezreal  Azir  Malzahar   
352  Blue      sup      Baolan  Invictus Gaming  Tahm Kench  Azir  Malzahar   

        ban3    ban4      ban5  gamelength  result  kills  deaths  assists  \
348  Camille  Illaoi  Vladimir        1590       1      3       1        3   
349  Camille  Illaoi  Vladimir        1590       1      3       0        5   
350  Camille  Illaoi  Vladimir        1590       1      3       1        8   
351  Camille  Illaoi  Vladimir        1590       1      3       1        7   
352  Camille  Illaoi  Vladimir        1590       1      0       1        9   

     teamkills  teamdeaths  doublekills  triplekills  quadrakills  pentakills  \
348         12           4          0.0          0.0          0.0         0.0   
349         12           4          0.0          0.0          0.0         0.0   
350         12           4          1.0          0.0          0.0         0.0   
351         12           4          0.0          0.0          0.0         0.0   
352         12           4          0.0          0.0          0.0         0.0   

     firstblood  firstbloodkill  firstbloodassist  firstbloodvictim  team kpm  \
348         1.0             0.0               1.0               0.0    0.4528   
349         1.0             1.0               0.0               0.0    0.4528   
350         1.0             0.0               1.0               0.0    0.4528   
351         1.0             0.0               1.0               0.0    0.4528   
352         1.0             0.0               1.0               0.0    0.4528   

       ckpm  firstdragon  dragons  opp_dragons  elementaldrakes  \
348  0.6038          NaN      NaN          NaN              NaN   
349  0.6038          NaN      NaN          NaN              NaN   
350  0.6038          NaN      NaN          NaN              NaN   
351  0.6038          NaN      NaN          NaN              NaN   
352  0.6038          NaN      NaN          NaN              NaN   

     opp_elementaldrakes  infernals  mountains  clouds  oceans  \
348                  NaN        NaN        NaN     NaN     NaN   
349                  NaN        NaN        NaN     NaN     NaN   
350                  NaN        NaN        NaN     NaN     NaN   
351                  NaN        NaN        NaN     NaN     NaN   
352                  NaN        NaN        NaN     NaN     NaN   

     dragons (type unknown)  elders  opp_elders  firstherald  heralds  \
348                     NaN     NaN         NaN          NaN      NaN   
349                     NaN     NaN         NaN          NaN      NaN   
350                     NaN     NaN         NaN          NaN      NaN   
351                     NaN   

In [5]:
lcs = df[df['league'].isin(['LCS', 'NA LCS'])]
lcs.shape

(7590, 117)

In [7]:
lcs = lcs.sort_values(by=['date', 'champion'])

In [9]:
lcs.head()

gameid datacompleteness  \
1701  TRLH1/1002440062         complete   
1692  TRLH1/1002440062         complete   
1697  TRLH1/1002440062         complete   
1693  TRLH1/1002440062         complete   
1700  TRLH1/1002440062         complete   

                                                    url  league  year   split  \
1701  http://matchhistory.na.leagueoflegends.com/en/...  NA LCS  2018  Spring   
1692  http://matchhistory.na.leagueoflegends.com/en/...  NA LCS  2018  Spring   
1697  http://matchhistory.na.leagueoflegends.com/en/...  NA LCS  2018  Spring   
1693  http://matchhistory.na.leagueoflegends.com/en/...  NA LCS  2018  Spring   
1700  http://matchhistory.na.leagueoflegends.com/en/...  NA LCS  2018  Spring   

      playoffs                date  game  patch  playerid  side position  \
1701         0 2018-01-20 20:50:09     1   8.01        10   Red      sup   
1692         0 2018-01-20 20:50:09     1   8.01         1  Blue      top   
1697         0 2018-01-20 20:50:09     1   8.01         6   Red      top   
1693         0 2018-01-20 20:50:09     1   8.01         2  Blue      jng   
1700         0 2018-01-20 20:50:09     1   8.01         9   Red      bot   

        player         team   champion     ban1        ban2    ban3    ban4  \
1701     Mithy          TSM    Alistar  Kog'Maw  Tahm Kench  Ezreal   Braum   
1692    Impact  Team Liquid  Gangplank     Ornn         Zoe    Azir  Thresh   
1697  Hauntzer          TSM       Gnar  Kog'Maw  Tahm Kench  Ezreal   Braum   
1693   Xmithie  Team Liquid     Gragas     Ornn         Zoe    Azir  Thresh   
1700      Zven          TSM    Kalista  Kog'Maw  Tahm Kench  Ezreal   Braum   

         ban5  gamelength  result  kills  deaths  assists  teamkills  \
1701     Shen        1687       0      0       3        1          1   
1692  Kha'Zix        1687       1      3       0        6         11   
1697     Shen        1687       0      0       3        0          1   
1693  Kha'Zix        1687       1      0       0        9         11   
1700     Shen        1687       0      0       2        0          1   

      teamdeaths  doublekills  triplekills  quadrakills  pentakills  \
1701          11          0.0          0.0          0.0         0.0   
1692           1          1.0          0.0          0.0         0.0   
1697          11          0.0          0.0          0.0         0.0   
1693           1          0.0          0.0          0.0         0.0   
1700          11          0.0          0.0          0.0         0.0   

      firstblood  firstbloodkill  firstbloodassist  firstbloodvictim  \
1701         0.0             0.0               0.0               0.0   
1692         1.0             0.0               1.0               0.0   
1697         0.0             0.0               0.0               1.0   
1693         0.0             0.0               0.0               0.0   
1700         0.0             0.0               0.0               0.0   

      team kpm    ckpm  firstdragon  dragons  opp_dragons  elementaldrakes  \
1701    0.0356  0.4268          NaN      NaN          NaN              NaN   
1692    0.3912  0.4268          NaN      NaN          NaN              NaN   
1697    0.0356  0.4268          NaN      NaN          NaN              NaN   
1693    0.3912  0.4268          NaN      NaN          NaN              NaN   
1700    0.0356  0.4268          NaN      NaN          NaN              NaN   

      opp_elementaldrakes  infernals  mountains  clouds  oceans  \
1701                  NaN        NaN        NaN     NaN     NaN   
1692                  NaN        NaN        NaN     NaN     NaN   
1697                  NaN        NaN        NaN     NaN     NaN   
1693                  NaN        NaN        NaN     NaN     NaN   
1700                  NaN        NaN        NaN     NaN     NaN   

      dragons (type unknown)  elders  opp_elders  firstherald  heralds  \
1701                     NaN     NaN         NaN          NaN      NaN   
1692                     

In [10]:
input_data = lcs[['gameid', 'date', 'champion', 'side', 'team', 'position', 'result']]
input_data.head()

gameid                date   champion  side         team  \
1701  TRLH1/1002440062 2018-01-20 20:50:09    Alistar   Red          TSM   
1692  TRLH1/1002440062 2018-01-20 20:50:09  Gangplank  Blue  Team Liquid   
1697  TRLH1/1002440062 2018-01-20 20:50:09       Gnar   Red          TSM   
1693  TRLH1/1002440062 2018-01-20 20:50:09     Gragas  Blue  Team Liquid   
1700  TRLH1/1002440062 2018-01-20 20:50:09    Kalista   Red          TSM   

     position  result  
1701      sup       0  
1692      top       1  
1697      top       0  
1693      jng       1  
1700      bot       0

In [11]:
columns = ['gameid', 'date', \
           'blue_team', \
           'blue_top_name', \
           'blue_jng_name', \
           'blue_mid_name', \
           'blue_bot_name', \
           'blue_sup_name',\
           'blue_team_result', \
           'red_team', \
           'red_top_name', \
           'red_jng_name', \
           'red_mid_name', \
           'red_bot_name', \
           'red_sup_name']

lcs_rating = pd.DataFrame(columns=columns)

## Trueskill

In [12]:
ts = trueskill.TrueSkill(draw_probability=0.0)

In [13]:
champion_ratings_dict = dict()
for i in input_data['champion'].unique():
    champion_ratings_dict[i] = ts.create_rating()
    
print(champion_ratings_dict)

{'Alistar': trueskill.Rating(mu=25.000, sigma=8.333), 'Gangplank': trueskill.Rating(mu=25.000, sigma=8.333), 'Gnar': trueskill.Rating(mu=25.000, sigma=8.333), 'Gragas': trueskill.Rating(mu=25.000, sigma=8.333), 'Kalista': trueskill.Rating(mu=25.000, sigma=8.333), 'Malzahar': trueskill.Rating(mu=25.000, sigma=8.333), 'Ryze': trueskill.Rating(mu=25.000, sigma=8.333), 'Shyvana': trueskill.Rating(mu=25.000, sigma=8.333), 'Taric': trueskill.Rating(mu=25.000, sigma=8.333), 'Tristana': trueskill.Rating(mu=25.000, sigma=8.333), 'Braum': trueskill.Rating(mu=25.000, sigma=8.333), "Cho'Gath": trueskill.Rating(mu=25.000, sigma=8.333), 'Ezreal': trueskill.Rating(mu=25.000, sigma=8.333), "Kog'Maw": trueskill.Rating(mu=25.000, sigma=8.333), 'Morgana': trueskill.Rating(mu=25.000, sigma=8.333), 'Rengar': trueskill.Rating(mu=25.000, sigma=8.333), 'Sejuani': trueskill.Rating(mu=25.000, sigma=8.333), 'Shen': trueskill.Rating(mu=25.000, sigma=8.333), 'Azir': trueskill.Rating(mu=25.000, sigma=8.333), 'Evely

In [14]:
### Lambda Apply w Position as Key
def setup_match(df, gameid, position):
    #blue players
    game_frame = df[df.gameid == gameid]
    blue_data = game_frame[game_frame.side == 'Blue']
    red_data = game_frame[game_frame.side == 'Red']

    blue_top_name = blue_data[blue_data.position == 'top'].champion.iloc[0]
    blue_jng_name = blue_data[blue_data.position == 'jng'].champion.iloc[0]
    blue_mid_name = blue_data[blue_data.position == 'mid'].champion.iloc[0]
    blue_bot_name = blue_data[blue_data.position == 'bot'].champion.iloc[0]
    blue_sup_name = blue_data[blue_data.position == 'sup'].champion.iloc[0]

    # red players
    red_top_name = red_data[red_data.position == 'top'].champion.iloc[0]
    red_jng_name = red_data[red_data.position == 'jng'].champion.iloc[0]
    red_mid_name = red_data[red_data.position == 'mid'].champion.iloc[0]
    red_bot_name = red_data[red_data.position == 'bot'].champion.iloc[0]
    red_sup_name = red_data[red_data.position == 'sup'].champion.iloc[0]

    # date, teams, result
    date = blue_data['date'].iloc[0]
    blue_team_result = blue_data['result'].iloc[0]
    blue_team = blue_data['team'].iloc[0]
    red_team = red_data['team'].iloc[0]
    # create pandas series for df
    match = pd.Series([gameid, date, \
                       blue_team, \
                       blue_top_name, \
                       blue_jng_name, \
                       blue_mid_name, \
                       blue_bot_name, \
                       blue_sup_name,\
                       blue_team_result, \
                       red_team, \
                       red_top_name, \
                       red_jng_name, \
                       red_mid_name, \
                       red_bot_name, \
                       red_sup_name])
    return match


lcs_rating[['gameid', 'date', \
           'blue_team', \
           'blue_top_name', \
           'blue_jng_name', \
           'blue_mid_name', \
           'blue_bot_name', \
           'blue_sup_name',\
           'blue_team_result', \
           'red_team', \
           'red_top_name', \
           'red_jng_name', \
           'red_mid_name', \
           'red_bot_name', \
           'red_sup_name']] = input_data.apply(lambda row: setup_match(input_data, row['gameid'], row['position']),axis=1)


In [15]:
lcs_rating.shape


(7590, 15)

In [16]:
lcs_rating = lcs_rating.reset_index(drop=True)
lcs_rating = lcs_rating.drop_duplicates(subset=['gameid'])
lcs_rating = lcs_rating.sort_values(by=['date'])

In [17]:
lcs_rating.head(20)

gameid                date             blue_team blue_top_name  \
0    TRLH1/1002440062 2018-01-20 20:50:09           Team Liquid     Gangplank   
10   TRLH1/1002440076 2018-01-20 22:38:20           100 Thieves          Shen   
20   TRLH1/1002440084 2018-01-21 00:14:20         Clutch Gaming     Gangplank   
30   TRLH1/1002440095 2018-01-21 01:20:39              Echo Fox        Lucian   
40   TRLH1/1002440106 2018-01-21 02:16:56  Counter Logic Gaming          Ornn   
50   TRLH1/1002440127 2018-01-21 21:49:32          OpTic Gaming          Ornn   
60   TRLH1/1002440132 2018-01-21 22:46:16              FlyQuest          Gnar   
70   TRLH1/1002440143 2018-01-22 00:00:28           100 Thieves          Gnar   
80   TRLH1/1002440150 2018-01-22 01:00:51      Golden Guardians     Gangplank   
90   TRLH1/1002440161 2018-01-22 02:02:56              Echo Fox          Gnar   
100  TRLH1/1002440178 2018-01-27 21:35:50         Clutch Gaming     Gangplank   
110  TRLH1/1002440188 2018-01-27 22:39:24                   TSM     Gangplank   
120  TRLH1/1002440193 2018-01-28 00:11:30                Cloud9         Jayce   
130  TRLH1/1002440202 2018-01-28 01:24:28           100 Thieves          Gnar   
140  TRLH1/1002440210 2018-01-28 02:37:26      Golden Guardians        Maokai   
150  TRLH1/1002440231 2018-01-28 21:44:50              Echo Fox     Gangplank   
160  TRLH1/1002440242 2018-01-28 23:04:25                Cloud9       Camille   
170  TRLH1/1002440256 2018-01-29 00:09:21          OpTic Gaming        Maokai   
180  TRLH1/1002440271 2018-01-29 01:11:36      Golden Guardians          Gnar   
190  TRLH1/1002440282 2018-01-29 02:18:47           Team Liquid     Gangplank   

    blue_jng_name blue_mid_name blue_bot_name blue_sup_name  blue_team_result  \
0          Gragas      Malzahar      Tristana         Taric                 1   
10        Sejuani          Ryze       Kog'Maw         Braum                 1   
20        Evelynn      Malzahar      Tristana          Ornn                 1   
30            Zac          Ryze       Kog'Maw          Shen                 1   
40         Rengar      Malzahar         Varus         Taric                 0   
50      Jarvan IV       Orianna       Kog'Maw         Taric                 0   
60            Zac          Azir       Kalista         Taric                 1   
70      Jarvan IV      Malzahar        Ezreal         Braum                 1   
80      Jarvan IV           Zoe         Varus         Taric                 0   
90            Zac          Azir        Ezreal         Taric                 1   
100           Zac          Azir         Varus       Morgana                 1   
110       Evelynn          Azir        Ezreal          Ornn                 1   
120           Jax          Ryze         Sivir       Alistar                 0   
130       Sejuani          Ryze       Kog'Maw         Braum                 1   
140        Rengar       Orianna       Kalista        Thresh                 0   
150           Zac           Zoe      Tristana       Alistar                 1   
160       Sejuani       Taliyah       Kalista         Taric                 1   
170       Kha'Zix        Syndra        Ezreal    Tahm Kench                 1   
180     Jarvan IV      Vladimir        Ezreal         Braum                 0   
190        Gragas      Malzahar      Tristana         Taric                 1   

                 red_team red_top_name red_jng_name red_mid_name red_bot_name  \
0                     TSM         Gnar      Shyvana         Ryze      Kalista   
10           OpTic Gaming     Cho'Gath       Rengar     Malzahar       Ezreal   
20       Golden Guardians     Vladimir      Kha'Zix         Azir      Kog'Maw   
30               FlyQuest    Gangplank    Jarvan IV   Cassiopeia        Varus   
40                 Cloud9    Gangplank      Sejuani       Zilean     Tristana   
50            Team Liquid         Gnar      Sejuani     Malzahar      Kalista   
60                    TSM     Cho'Gath      

In [18]:
analyzed_gameids = {}

def win_probability(team1, team2, trueskill_global_env):
    delta_mu = sum(r.mu for r in team1) - sum(r.mu for r in team2)
    sum_sigma = sum(r.sigma ** 2 for r in itertools.chain(team1, team2))
    size = len(team1) + len(team2)
    denom = math.sqrt(size * (trueskill_global_env.beta ** 2 ) + sum_sigma)
    return trueskill_global_env.cdf(delta_mu / denom)

def update_trueskill(rating_dict, gameid_dict, gameid, blue_team_result, \
                        blue_top_name, \
                        blue_jng_name, \
                        blue_mid_name, \
                        blue_bot_name, \
                        blue_sup_name,\
                        red_top_name, \
                        red_jng_name, \
                        red_mid_name, \
                        red_bot_name, \
                        red_sup_name):
    rating_groups = [(rating_dict[blue_top_name], \
                      rating_dict[blue_jng_name], \
                      rating_dict[blue_mid_name], \
                      rating_dict[blue_bot_name], \
                      rating_dict[blue_sup_name]), \
                     (rating_dict[red_top_name], \
                      rating_dict[red_jng_name], \
                      rating_dict[red_mid_name], \
                      rating_dict[red_bot_name], \
                      rating_dict[red_sup_name])]
    blue_mu = rating_groups[0][0].mu
    blue_sigma = rating_groups[0][0].sigma
    red_mu = rating_groups[1][0].mu
    red_sigma = rating_groups[1][0].sigma
    blue_team_win_prob = win_probability(rating_groups[0], rating_groups[1], ts)
    # Get Mu by position
    blue_top_mu = rating_dict[blue_top_name].mu
    blue_jng_mu = rating_dict[blue_jng_name].mu
    blue_mid_mu = rating_dict[blue_mid_name].mu
    blue_bot_mu = rating_dict[blue_bot_name].mu
    blue_sup_mu = rating_dict[blue_sup_name].mu
    red_top_mu = rating_dict[red_top_name].mu
    red_jng_mu = rating_dict[red_jng_name].mu
    red_mid_mu = rating_dict[red_mid_name].mu
    red_bot_mu = rating_dict[red_bot_name].mu
    red_sup_mu = rating_dict[red_sup_name].mu
    # Get Sigma by position
    blue_top_sigma = rating_dict[blue_top_name].sigma
    blue_jng_sigma = rating_dict[blue_jng_name].sigma
    blue_mid_sigma = rating_dict[blue_mid_name].sigma
    blue_bot_sigma = rating_dict[blue_bot_name].sigma
    blue_sup_sigma = rating_dict[blue_sup_name].sigma
    red_top_sigma = rating_dict[red_top_name].sigma
    red_jng_sigma = rating_dict[red_jng_name].sigma
    red_mid_sigma = rating_dict[red_mid_name].sigma
    red_bot_sigma = rating_dict[red_bot_name].sigma
    red_sup_sigma = rating_dict[red_sup_name].sigma
    # Update ratings
    if blue_team_result == 1:
        rated_rating_groups = ts.rate(rating_groups, ranks=[0, 1]) # for ranks 0 is winner 
    else:
        rated_rating_groups = ts.rate(rating_groups, ranks=[1, 0])
    # Return values for new columns
    ts_update = pd.Series([blue_team_win_prob, blue_mu, blue_sigma, red_mu, red_sigma, \
           blue_top_mu, blue_top_sigma, \
           blue_jng_mu, blue_jng_sigma, \
           blue_mid_mu, blue_mid_sigma, \
           blue_bot_mu, blue_bot_sigma, \
           blue_sup_mu, blue_sup_sigma, \
           red_top_mu, red_top_sigma, \
           red_jng_mu, red_jng_sigma, \
           red_mid_mu, red_mid_sigma, \
           red_bot_mu, red_bot_sigma, \
           red_sup_mu, red_sup_sigma])
    # update the rating dictionary
    rating_dict[blue_top_name] = rated_rating_groups[0][0]
    rating_dict[blue_jng_name] = rated_rating_groups[0][1]
    rating_dict[blue_mid_name] = rated_rating_groups[0][2]
    rating_dict[blue_bot_name] = rated_rating_groups[0][3]
    rating_dict[blue_sup_name] = rated_rating_groups[0][4]
    rating_dict[red_top_name] = rated_rating_groups[1][0]
    rating_dict[red_jng_name] = rated_rating_groups[1][1]
    rating_dict[red_mid_name] = rated_rating_groups[1][2]
    rating_dict[red_bot_name] = rated_rating_groups[1][3]
    rating_dict[red_sup_name] = rated_rating_groups[1][4]
    # determine if use old data because of repeat
    if gameid in gameid_dict:
        return gameid_dict[gameid]
    else:
        gameid_dict[gameid] = ts_update
        return ts_update

lcs_rating[['blue_team_win_prob', 'blue_mu', 'blue_sigma', 'red_mu', 'red_sigma',\
           'blue_top_mu', 'blue_top_sigma', \
           'blue_jng_mu', 'blue_jng_sigma', \
           'blue_mid_mu', 'blue_mid_sigma', \
           'blue_bot_mu', 'blue_bot_sigma', \
           'blue_sup_mu', 'blue_sup_sigma', \
           'red_top_mu', 'red_top_sigma', \
           'red_jng_mu', 'red_jng_sigma', \
           'red_mid_mu', 'red_mid_sigma', \
           'red_bot_mu', 'red_bot_sigma', \
           'red_sup_mu', 'red_sup_sigma']] = lcs_rating.apply(lambda row: update_trueskill(champion_ratings_dict, analyzed_gameids, row['gameid'], row['blue_team_result'], \
                        row['blue_top_name'], \
                        row['blue_jng_name'], \
                        row['blue_mid_name'], \
                        row['blue_bot_name'], \
                        row['blue_sup_name'],\
                        row['red_top_name'], \
                        row['red_jng_name'], \
                        row['red_mid_name'], \
                        row['red_bot_name'], \
                        row['red_sup_name']), axis=1)

In [19]:
lcs_rating.shape

(759, 40)

In [20]:
lcs_rating = lcs_rating.reset_index(drop=True)
lcs_rating.head(20)

gameid                date             blue_team blue_top_name  \
0   TRLH1/1002440062 2018-01-20 20:50:09           Team Liquid     Gangplank   
1   TRLH1/1002440076 2018-01-20 22:38:20           100 Thieves          Shen   
2   TRLH1/1002440084 2018-01-21 00:14:20         Clutch Gaming     Gangplank   
3   TRLH1/1002440095 2018-01-21 01:20:39              Echo Fox        Lucian   
4   TRLH1/1002440106 2018-01-21 02:16:56  Counter Logic Gaming          Ornn   
5   TRLH1/1002440127 2018-01-21 21:49:32          OpTic Gaming          Ornn   
6   TRLH1/1002440132 2018-01-21 22:46:16              FlyQuest          Gnar   
7   TRLH1/1002440143 2018-01-22 00:00:28           100 Thieves          Gnar   
8   TRLH1/1002440150 2018-01-22 01:00:51      Golden Guardians     Gangplank   
9   TRLH1/1002440161 2018-01-22 02:02:56              Echo Fox          Gnar   
10  TRLH1/1002440178 2018-01-27 21:35:50         Clutch Gaming     Gangplank   
11  TRLH1/1002440188 2018-01-27 22:39:24                   TSM     Gangplank   
12  TRLH1/1002440193 2018-01-28 00:11:30                Cloud9         Jayce   
13  TRLH1/1002440202 2018-01-28 01:24:28           100 Thieves          Gnar   
14  TRLH1/1002440210 2018-01-28 02:37:26      Golden Guardians        Maokai   
15  TRLH1/1002440231 2018-01-28 21:44:50              Echo Fox     Gangplank   
16  TRLH1/1002440242 2018-01-28 23:04:25                Cloud9       Camille   
17  TRLH1/1002440256 2018-01-29 00:09:21          OpTic Gaming        Maokai   
18  TRLH1/1002440271 2018-01-29 01:11:36      Golden Guardians          Gnar   
19  TRLH1/1002440282 2018-01-29 02:18:47           Team Liquid     Gangplank   

   blue_jng_name blue_mid_name blue_bot_name blue_sup_name  blue_team_result  \
0         Gragas      Malzahar      Tristana         Taric                 1   
1        Sejuani          Ryze       Kog'Maw         Braum                 1   
2        Evelynn      Malzahar      Tristana          Ornn                 1   
3            Zac          Ryze       Kog'Maw          Shen                 1   
4         Rengar      Malzahar         Varus         Taric                 0   
5      Jarvan IV       Orianna       Kog'Maw         Taric                 0   
6            Zac          Azir       Kalista         Taric                 1   
7      Jarvan IV      Malzahar        Ezreal         Braum                 1   
8      Jarvan IV           Zoe         Varus         Taric                 0   
9            Zac          Azir        Ezreal         Taric                 1   
10           Zac          Azir         Varus       Morgana                 1   
11       Evelynn          Azir        Ezreal          Ornn                 1   
12           Jax          Ryze         Sivir       Alistar                 0   
13       Sejuani          Ryze       Kog'Maw         Braum                 1   
14        Rengar       Orianna       Kalista        Thresh                 0   
15           Zac           Zoe      Tristana       Alistar                 1   
16       Sejuani       Taliyah       Kalista         Taric                 1   
17       Kha'Zix        Syndra        Ezreal    Tahm Kench                 1   
18     Jarvan IV      Vladimir        Ezreal         Braum                 0   
19        Gragas      Malzahar      Tristana         Taric                 1   

                red_team red_top_name red_jng_name red_mid_name red_bot_name  \
0                    TSM         Gnar      Shyvana         Ryze      Kalista   
1           OpTic Gaming     Cho'Gath       Rengar     Malzahar       Ezreal   
2       Golden Guardians     Vladimir      Kha'Zix         Azir      Kog'Maw   
3               FlyQuest    Gangplank    Jarvan IV   Cassiopeia        Varus   
4                 Cloud9    Gangplank      Sejuani       Zilean     Tristana   
5            Team Liquid         Gnar      Sejuani     Malzahar      Kalista   
6                    TSM     Cho'Gath      Kha'Zix       Xerath      Kog'Maw   
7   Counter 

### Calculate Future Winner from Input Teams

Team1 and Team2 input tuples for win_probability function  

(trueskill.Rating(mu=25.000, sigma=8.333), trueskill.Rating(mu=25.000, sigma=8.333), trueskill.Rating(mu=25.000, sigma=8.333), trueskill.Rating(mu=25.000, sigma=8.333), trueskill.Rating(mu=25.000, sigma=8.333))  

(trueskill.Rating(mu=25.000, sigma=8.333), trueskill.Rating(mu=25.000, sigma=8.333), trueskill.Rating(mu=25.000, sigma=8.333), trueskill.Rating(mu=25.000, sigma=8.333), trueskill.Rating(mu=25.000, sigma=8.333))

In [21]:
print(champion_ratings_dict)

{'Alistar': trueskill.Rating(mu=25.407, sigma=2.258), 'Gangplank': trueskill.Rating(mu=30.971, sigma=2.048), 'Gnar': trueskill.Rating(mu=20.287, sigma=2.894), 'Gragas': trueskill.Rating(mu=27.609, sigma=2.134), 'Kalista': trueskill.Rating(mu=24.545, sigma=2.556), 'Malzahar': trueskill.Rating(mu=36.047, sigma=4.561), 'Ryze': trueskill.Rating(mu=22.751, sigma=2.372), 'Shyvana': trueskill.Rating(mu=22.127, sigma=7.966), 'Taric': trueskill.Rating(mu=25.646, sigma=3.575), 'Tristana': trueskill.Rating(mu=21.367, sigma=3.382), 'Braum': trueskill.Rating(mu=25.672, sigma=1.642), "Cho'Gath": trueskill.Rating(mu=23.701, sigma=3.322), 'Ezreal': trueskill.Rating(mu=24.512, sigma=1.671), "Kog'Maw": trueskill.Rating(mu=24.995, sigma=4.065), 'Morgana': trueskill.Rating(mu=25.443, sigma=2.875), 'Rengar': trueskill.Rating(mu=18.447, sigma=7.384), 'Sejuani': trueskill.Rating(mu=24.246, sigma=1.801), 'Shen': trueskill.Rating(mu=27.755, sigma=2.804), 'Azir': trueskill.Rating(mu=24.470, sigma=2.031), 'Evely

TSM = ['Huni', 'Spica', 'PowerOfEvil', 'Lost', 'SwordArt']
FLY = ['Licorice', 'Josedeodo', 'Palafox', 'Johnsun', 'Diamond']
TSM_RATINGS = []
FLY_RATINGS = []

for player in TSM:
    TSM_RATINGS.append(player_ratings_dict[player])

for player in FLY:
    FLY_RATINGS.append(player_ratings_dict[player])
    
TSM_TUPLE = tuple(TSM_RATINGS)
FLY_TUPLE = tuple(FLY_RATINGS)

team_1_win_prob = win_probability(TSM_TUPLE, FLY_TUPLE, ts)
print('TSM Win Probability :' + str(team_1_win_prob))

In [22]:
#print(analyzed_gameids)

In [25]:
#lcs_rating.to_csv('./lcs_champion_trueskill_2018-2021.csv', index=False)